### PR TITLE
Finetune only on target labels

### PR DESCRIPTION
**Pull request by sriniiyer**
**Description:**
This diff allows us to backpropagate losses only on designated target tokens rather than all the input tokens. It introduces a new jsonl data format for this, where "src" and "tgt" keys are specified. To use this, use the streaming_finetune_language_modeling task. The task concatenates the src and the tgt, and is trained to produce this autoregressively, but only takes losses on the tgt tokens into account. This is particularly useful during fine-tuning.

**Test plan**
1. Does not change existing pre-training i.e. tested existing pre-training on 125M model and ppl for the first 5 updates were exactly the same i.e. unchanged.

2. Works successfully on src-tgt-format without crashing and performs competitively on copa from superglue.

### CODE REVIEW

File: metaseq/tasks/streaming_finetune_language_modeling.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            # We generate blocks with one extra token, so that we have a target
+            # for the final input token. This results in slight data loss.
+            block_size=self.args.tokens_per_sample + 1,
+            break_mode=self.args.sample_break_mode,
+            # we drop the remainder block during training
+            drop_last=(split == "train"),
+            padding_idx=self.source_dictionary.pad(),
+            # 1284 is a randomly-generated offset to decouple the seed used here
+            # from the seed used above in StreamingShuffleDataset
+            seed=1284 + self.args.seed,
```
Comment (suchenzang) [inline metaseq/tasks/streaming_finetune_language_modeling.py:107]: Nit: Would be good to keep track of all the RNG offsets in our codebase and where collisions may occur (or add a TODO pointing to https://github.com/facebookresearch/metaseq/issues/65)

File: metaseq/data/streaming_src_tgt_dataset.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            self.dataset,
+            self.block_size,
+            self.drop_last,
+            self.padding_idx,
+        )
+
+        if self.seed is not None:
+            # add a random offset (2273) to the given seed to decouple this RNG
+            # from any other RNG instances elsewhere
+            rng = np.random.default_rng(2273 + self.seed)
```
Comment (suchenzang) [inline metaseq/data/streaming_src_tgt_dataset.py:89]: This looks to be the same RNG offset copied from StreamingTokenBlockDataset. Is that intentional?

File: metaseq/data/streaming_src_tgt_dataset.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+import numpy as np
+import torch
+
+
+class StreamingSrcTgtDataset(torch.utils.data.IterableDataset):
```
Comment (suchenzang) [inline metaseq/data/streaming_src_tgt_dataset.py:12]: Is there any way for this class to extend StreamingTokenBlockDataset instead?

File: metaseq/tasks/streaming_finetune_language_modeling.py (hunk 1 above)
Comment (sriniiyer) [reply metaseq/tasks/streaming_finetune_language_modeling.py:107]: Will add TODO for now.

File: metaseq/data/streaming_src_tgt_dataset.py (hunk 2 above)
Comment (suchenzang) [reply metaseq/data/streaming_src_tgt_dataset.py:89]: Ah it just needs to be something distinct for decoupling.  Can you pick another random offset to be safe? Opened https://github.com/facebookresearch/metaseq/issues/65 to tackle this more broadly.

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py [hunk 4]
```diff
@@ -103,17 +107,9 @@ def reduce_metrics(logging_outputs) -> None:
         metrics.log_scalar(
             "loss", loss_sum / sample_size / math.log(2), sample_size, round=3
         )
-        if sample_size != ntokens:
-            metrics.log_scalar(
```
Comment (suchenzang) [inline metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:107]: Why did this get taken out?

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py [hunk 5]
```diff
@@ -46,7 +46,11 @@ def forward(self, model, sample, reduce=True):
         if has_pad:
             loss = loss * (target != self.padding_idx)
         loss = loss.sum()
-        sample_size = sample["ntokens"]
+        sample_size = (
+            sample["ntokens_target"]
```
Comment (suchenzang) [inline metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:50]: Nit: add a comment for what this logic is for / where `ntokens_target` is present vs `ntokens`?

File: metaseq/tasks/streaming_language_modeling.py [hunk 6]
```diff
@@ -266,16 +279,7 @@ def load_dataset(self, split: str, epoch=1, combine=False, **kwargs):
...
-        for shard_id in os.listdir(os.path.join(self.args.data, split)):
-            assert (
-                int(shard_id) not in shards
-            ), f"shard id: {shard_id} not in shards: {shards}"
-            shards[int(shard_id)] = shard_id
-        assert min(shards.keys()) == 0
-        assert max(shards.keys()) == len(shards) - 1
-
-        cur_shard_str = shards[(epoch - 1) % len(shards)]
+        cur_shard_str = self.get_shard_str(epoch, split)
```
Comment (suchenzang) [inline metaseq/tasks/streaming_language_modeling.py:282]: To confirm: this logic remains unchanged, correct?  The main change here is to move it into a separate function?

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py (hunk 4 above)
Comment (sriniiyer) [reply metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:107]: Hmm, so in the original code, sample_size is always equal to ntokens, so this is never executed and it doesn't change anything

In the new code, we aren't interested in nll_loss, it's some code from the MT days I think.

File: metaseq/tasks/streaming_language_modeling.py (hunk 6 above)
Comment (sriniiyer) [reply metaseq/tasks/streaming_language_modeling.py:282]: Correct. Just to reuse this code in the derived class.

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py (hunk 5 above)
Comment (sriniiyer) [reply metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:50]: Will add

File: metaseq/data/streaming_src_tgt_dataset.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    """Packs multiple examples together in a block"""
+    cur_src_block = []
+    cur_src_block_ids = []
+    cur_tgt_block = []
+    cur_block_remain = block_size
+    for idx, (src, tgt) in enumerate(iterable):
+        if src.numel() > block_size:
+            # truncate right side
+            src = src[:block_size]
+            tgt = tgt[:block_size]
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply metaseq/data/streaming_src_tgt_dataset.py:129]: also cc @klshuster, this will bite us

File: metaseq/tasks/streaming_finetune_language_modeling.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
```
Comment (stephenroller) [inline metaseq/tasks/streaming_finetune_language_modeling.py:1]: let's please leave an issue saying to have someone possibly refactor this in the future

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sriniiyer) [reply metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:107]: But this code path is never executed right now

File: metaseq/tasks/streaming_language_modeling.py [hunk 9]
```diff
@@ -242,6 +242,19 @@ def _alpha_sampling(self, datasets, corpora, epoch=1):
         )
         return resampled_datasets
 
+    def get_shard_str(self, epoch, split):
```
[the remark this answers is not in the document]
Comment (sriniiyer) [reply metaseq/tasks/streaming_language_modeling.py:245]: Wanted to re-use this in the derived class. Is it ok to duplicate this code in the derived class then? Any other better way?

File: metaseq/data/streaming_src_tgt_dataset.py (hunk 2 above)
Comment (sriniiyer) [reply metaseq/data/streaming_src_tgt_dataset.py:89]: Changing to self.rng. We dont need to change the number since this dataset is used independently of token_block_dataset.

File: metaseq/tasks/streaming_finetune_language_modeling.py (hunk 8 above)
Comment (sriniiyer) [reply metaseq/tasks/streaming_finetune_language_modeling.py:1]: Will add an issue once this is landed

File: metaseq/tasks/streaming_finetune_language_modeling.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+            /path/to/data/train/01/foo.jsonl
+            /path/to/data/train/01/bar.jsonl
+            /path/to/data/valid/00/foo.jsonl
+            /path/to/data/valid/00/bar.jsonl
+
+        In this example, we have two "shards" of training data, which will be
+        iterated over in epochs 1 and 2, respectively. Subsequent epochs will
+        cycle back over the same data. We also have two different data sources
+        in each shard (foo and bar), which will be combined and shuffled.
+
```
[the remark this answers is not in the document]
Comment (sriniiyer) [reply metaseq/tasks/streaming_finetune_language_modeling.py:62]: Done

File: metaseq/tasks/streaming_language_modeling.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sriniiyer) [reply metaseq/tasks/streaming_language_modeling.py:245]: @stephenroller Does this work?

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (stephenroller) [reply metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:107]: Okay i guess we still have loss. good with me.

**Review by stephenroller — APPROVED** [review]
